mshv-ioctls/src/ioctls/vcpu.rs

```
@@ -947,6 +947,63 @@ impl VcpuFd {
 
         Ok(ret_regs)
     }
+    #[cfg(target_arch = "x86_64")]
+    ///
+    /// Register override CPUID values for one leaf.
+    ///
+    pub fn register_intercept_result_cpuid_entry(&self, entry: &hv_cpuid_entry) -> Result<()> {
+        let mut mshv_cpuid = hv_register_x64_cpuid_result_parameters {
+            input: hv_register_x64_cpuid_result_parameters__bindgen_ty_1 {
+                eax: entry.function,
+                ecx: 0,
+                subleaf_specific: 0,
+                always_override: 1,
+                padding: 0,
+            },
+            result: hv_register_x64_cpuid_result_parameters__bindgen_ty_2 {
+                eax: entry.eax,
+                eax_mask: entry.eax,
+                ebx: entry.ebx,
+                ebx_mask: entry.ebx,
+                ecx: entry.ecx,
+                ecx_mask: entry.ecx,
+                edx: entry.edx,
+                edx_mask: entry.edx,
+            },
+        };
+        let args = mshv_register_intercept_result {
+            intercept_type: hv_intercept_type_HV_INTERCEPT_TYPE_X64_CPUID,
+            parameters: hv_register_intercept_result_parameters { cpuid: mshv_cpuid },
+        };
+        let ret = unsafe { ioctl_with_ref(self, MSHV_VP_REGISTER_INTERCEPT_RESULT(), &args) };
+        if ret != 0 {
+            return Err(errno::Error::last());
+        }
+
+        Ok(())
+    }
+    ///
+    /// Extend CPUID values delivered by hypervisor.
+    ///
+    #[cfg(target_arch = "x86_64")]
+    pub fn register_intercept_result_cpuid(&self, cpuid: &CpuId) -> Result<()> {
+        let mut ret = Ok(());
+
+        for entry in cpuid.as_slice().iter() {
+            let eret = self.register_intercept_result_cpuid_entry(entry);
+            if eret.is_err() {
+                if ret.is_ok() {
+                    ret = eret;
+                }
+                eprintln!(
```

This is not suitable for a library. You should return the error to its caller.

---

Some provide stdio hooks, some use the default. Do we have an infrastructure to bind the stdio to consumer? Or any other logging mechanism to hook into? Otherwise probably no better solution than just remove the print then.

Thanks

---

We cannot make any assumption here so please just remove the print. I don't know any well-known library that prints.

If you return an `Err` here the callers will deal with it as they see fit. You don't need to worry about logging anymore.